## 1. Idea

## 2. Initial exploration

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

#### Reading the DataFrame from the .csv file

In [7]:
df = pd.read_csv('train.csv')
dfTest = pd.read_csv('test.csv')

In [4]:
sns.set_style('whitegrid')

## 3. Introducing new features, gender/class subframes and function Difference

### 3.1. New features

In [64]:
df['LastName']=df['Name'].apply(lambda s: s.split()[0])

In [69]:
df['LastName'].value_counts().head(10)

Andersson,    9
Sage,         7
Johnson,      6
Carter,       6
Goodwin,      6
Panula,       6
Skoog,        6
Rice,         5
Lefebre,      4
Vander        4
Name: LastName, dtype: int64

In [71]:
df[df['LastName']=='Lefebre,']

PassengerId  Survived  Pclass                           Name     Sex  \
176          177         0       3  Lefebre, Master. Henry Forbes    male   
229          230         0       3        Lefebre, Miss. Mathilde  female   
409          410         0       3             Lefebre, Miss. Ida  female   
485          486         0       3         Lefebre, Miss. Jeannie  female   

     Age  SibSp  Parch Ticket     Fare Cabin Embarked CabType    Title  \
176  NaN      3      1   4133  25.4667   NaN        S       n  Master.   
229  NaN      3      1   4133  25.4667   NaN        S       n    Miss.   
409  NaN      3      1   4133  25.4667   NaN        S       n    Miss.   
485  NaN      3      1   4133  25.4667   NaN        S       n    Miss.   

     LastName  
176  Lefebre,  
229  Lefebre,  
409  Lefebre,  
485  Lefebre,

In [72]:
df[df['LastName']=='Skoog,']

PassengerId  Survived  Pclass  \
63            64         0       3   
167          168         0       3   
360          361         0       3   
634          635         0       3   
642          643         0       3   
819          820         0       3   

                                                Name     Sex   Age  SibSp  \
63                             Skoog, Master. Harald    male   4.0      3   
167  Skoog, Mrs. William (Anna Bernhardina Karlsson)  female  45.0      1   
360                               Skoog, Mr. Wilhelm    male  40.0      1   
634                               Skoog, Miss. Mabel  female   9.0      3   
642                    Skoog, Miss. Margit Elizabeth  female   2.0      3   
819                     Skoog, Master. Karl Thorsten    male  10.0      3   

     Parch  Ticket  Fare Cabin Embarked CabType    Title LastName  
63       2  347088  27.9   NaN        S       n  Master.   Skoog,  
167      4  347088  27.9   NaN        S       n     Mrs.   Skoog,  
360      4  347088  27.9   NaN        S       n      Mr.   Skoog,  
634      2  347088  27.9   NaN        S       n    Miss.   Skoog,  
642      2  347088  27.9   NaN        S       n    Miss.   Skoog,  
819      2  347088  27.9   NaN        S       n  Master.   Skoog,

## Search for people in the test sample that have companions in the train sample

In [8]:
df['PassengerId'].count()

891

In [12]:
dfTest['PassengerId'].count()

418

In [54]:
arr_tickets=[]
for i in range(418):       
    ticket1 = dfTest.loc[i]['Ticket']
    for j in range(891):
        ticket2 = df.loc[j]['Ticket']
        if (ticket1==ticket2):
            arr_tickets.append(ticket1)
            continue

In [59]:
arr_tickets_set=set(arr_tickets)
arr_tickets_list=list(arr_tickets_set)

In [61]:
np_arr = np.array(arr_tickets_list)
np_arr.size

115

### Searching for Female in the 3rd class among passengers that have cross-tickets

In [44]:
dfTest_F_cl3 = dfTest[(dfTest['Sex']=='female') & (dfTest['Pclass']==3)]
dfTest_F_cl3=dfTest_F_cl3.reset_index()
#dfTest_F_cl3.drop(['index'],inplace=True)
dfTest_F_cl3.head()

index  PassengerId  Pclass                                          Name  \
0      1          893       3              Wilkes, Mrs. James (Ellen Needs)   
1      4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
2      6          898       3                          Connolly, Miss. Kate   
3      8          900       3     Abrahim, Mrs. Joseph (Sophie Halaut Easu)   
4     18          910       3                  Ilmakangas, Miss. Ida Livija   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  female  47.0      1      0            363272   7.0000   NaN        S  
1  female  22.0      1      1           3101298  12.2875   NaN        S  
2  female  30.0      0      0            330972   7.6292   NaN        Q  
3  female  18.0      0      0              2657   7.2292   NaN        C  
4  female  27.0      1      0  STON/O2. 3101270   7.9250   NaN        S

In [37]:
dfTest_F_cl3['PassengerId'].count()

72

In [65]:
arr_ticks_F_cl3=[]
for i in range(72):
    ticket1 = dfTest_F_cl3.loc[i]['Ticket']
    for k in range(np_arr.size):
        if (ticket1==arr_tickets_list[k]): 
            arr_ticks_F_cl3.append(ticket1)
            continue


In [68]:
arr_ticks_F_cl3_set = set(arr_ticks_F_cl3)
arr_ticks_F_cl3_list = list(arr_ticks_F_cl3_set)

In [70]:
np.array(arr_ticks_F_cl3_list).size

23

- - -
#### Check ticket [0] in the list

In [76]:
dfTest[dfTest['Ticket']==arr_ticks_F_cl3_list[0]]

PassengerId  Pclass                            Name     Sex   Age  SibSp  \
139         1031       3  Goodwin, Mr. Charles Frederick    male  40.0      1   
140         1032       3     Goodwin, Miss. Jessie Allis  female  10.0      5   

     Parch   Ticket  Fare Cabin Embarked  
139      6  CA 2144  46.9   NaN        S  
140      2  CA 2144  46.9   NaN        S

In [72]:
df[df['Ticket']==arr_ticks_F_cl3_list[0]]

PassengerId  Survived  Pclass                                     Name  \
59            60         0       3       Goodwin, Master. William Frederick   
71            72         0       3               Goodwin, Miss. Lillian Amy   
386          387         0       3          Goodwin, Master. Sidney Leonard   
480          481         0       3           Goodwin, Master. Harold Victor   
678          679         0       3  Goodwin, Mrs. Frederick (Augusta Tyler)   
683          684         0       3              Goodwin, Mr. Charles Edward   

        Sex   Age  SibSp  Parch   Ticket  Fare Cabin Embarked  
59     male  11.0      5      2  CA 2144  46.9   NaN        S  
71   female  16.0      5      2  CA 2144  46.9   NaN        S  
386    male   1.0      5      2  CA 2144  46.9   NaN        S  
480    male   9.0      5      2  CA 2144  46.9   NaN        S  
678  female  43.0      1      6  CA 2144  46.9   NaN        S  
683    male  14.0      5      2  CA 2144  46.9   NaN        S

#### So, we have 2 passengers in the test set and 6 passengers in the training set, all in the same party. Passengers in the training set, males and females, grown-ups and little ones, all died. It's very likely that Mr. Charles Frederick and Miss. Jessie Allis shared their fate. Altough our original classifier would mark them both as died anyways.
- - -
#### What about ticket [1]?

In [75]:
dfTest[dfTest['Ticket']==arr_ticks_F_cl3_list[1]]

PassengerId  Pclass                                          Name     Sex  \
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
4  22.0      1      1  3101298  12.2875   NaN        S

In [74]:
df[df['Ticket']==arr_ticks_F_cl3_list[1]]

PassengerId  Survived  Pclass                      Name     Sex  Age  \
479          480         1       3  Hirvonen, Miss. Hildur E  female  2.0   

     SibSp  Parch   Ticket     Fare Cabin Embarked  
479      0      1  3101298  12.2875   NaN        S

#### I don't know where is Helga's spouse/sibling, but she looks very much like a mother of little Miss. Hildur E from our training set that managed to survive. It's very likely that the mother from the test set also survived. So, we label Passenger 896 as survived while our original classifier would have label her 'dead'.
- - -
#### And [2]:

In [77]:
dfTest[dfTest['Ticket']==arr_ticks_F_cl3_list[2]]

PassengerId  Pclass                      Name     Sex   Age  SibSp  \
125         1017       3  Cribb, Miss. Laura Alice  female  17.0      0   

     Parch  Ticket  Fare Cabin Embarked  
125      1  371362  16.1   NaN        S

In [78]:
df[df['Ticket']==arr_ticks_F_cl3_list[2]]

PassengerId  Survived  Pclass                      Name   Sex   Age  \
160          161         0       3  Cribb, Mr. John Hatfield  male  44.0   

     SibSp  Parch  Ticket  Fare Cabin Embarked  
160      0      1  371362  16.1   NaN        S

#### Miss. Laura Alice is 17 year old female and we know that her father from out training set died. As a female, she would have priority to survive, so this information doesn't help us. We mark her as 'died' according with our original classifier.
- - -

## 6. Prepare the submition file

In [48]:
dfNew=pd.DataFrame(columns=('PassengerId', 'Survived'))

for i in range(418):
    surv=0       
    PassId=dfTest.loc[i]['PassengerId']
    if (dfTest.loc[i]['Sex']=='female'): 
        if (dfTest.loc[i]['Pclass']==1): surv=1
        if (dfTest.loc[i]['Pclass']==2): surv=1
        if (dfTest.loc[i]['Pclass']==3): 
            if (dfTest.loc[i]['Embarked']=='Q'): surv=1
                
    if (dfTest.loc[i]['Sex']=='male'):
        if (dfTest.loc[i]['Pclass']==1): 
            if (dfTest.loc[i]['Age']<14): surv=1
        if (dfTest.loc[i]['Pclass']==2):
            if (dfTest.loc[i]['Age']<14): surv=1
            
    dfNew.loc[i] = pd.Series({'PassengerId':PassId,'Survived':surv})

dfNew.to_csv('submit1.csv',index=False)
# 0.77033

In [47]:
dfNew=pd.DataFrame(columns=('PassengerId', 'Survived'))

for i in range(418):
    surv=0       
    PassId=dfTest.loc[i]['PassengerId']
    if (dfTest.loc[i]['Sex']=='female'): 
        if (dfTest.loc[i]['Pclass']==1): surv=1
        if (dfTest.loc[i]['Pclass']==2): surv=1
        if (dfTest.loc[i]['Pclass']==3): 
            if (dfTest.loc[i]['Embarked']=='S'): pass
            elif (dfTest.loc[i]['Parch']>1): pass
            elif (dfTest.loc[i]['SibSp']>1): pass
            elif (dfTest.loc[i]['Age']>40): pass
            else: surv=0
            
                
    if (dfTest.loc[i]['Sex']=='male'):
        if (dfTest.loc[i]['Pclass']==1): 
            if (dfTest.loc[i]['Age']<14): surv=1
        if (dfTest.loc[i]['Pclass']==2):
            if (dfTest.loc[i]['Age']<14): surv=1
            
    dfNew.loc[i] = pd.Series({'PassengerId':PassId,'Survived':surv})

dfNew.to_csv('submit2.csv',index=False)
# 0.76555